pkg/util/retry/retry_test.go

```diff
@@ -140,26 +142,148 @@ func TestRetryNextCh(t *testing.T) {
 }
 
 func TestRetryWithMaxAttempts(t *testing.T) {
-	opts := Options{
-		InitialBackoff: time.Microsecond * 10,
-		MaxBackoff:     time.Second,
-		Multiplier:     2,
-		MaxRetries:     1,
-	}
-
-	attempts := 0
-	const maxAttempts = 3
 	expectedErr := errors.New("placeholder")
-	errFn := func() error {
+	attempts := 0
+	errWithAttemptsCounterFunc := func() error {
 		attempts++
 		return expectedErr
 	}
+	errorUntilAttemptNumFunc := func(until int) func() error {
+		return func() error {
+			attempts++
+			if attempts == until {
+				return nil
+			}
+			return expectedErr
+		}
+	}
+	cancelCtx, cancelCtxFunc := context.WithCancel(context.Background())
+	closeCh := make(chan struct{})
+
+	testCases := []struct {
+		desc string
+
+		ctx                    context.Context
+		opts                   Options
+		preWithMaxAttemptsFunc func()
+		retryFunc              func() error
+		maxAttempts            int
+
+		// Due to channel races with select, we can allow a range of number of attempts.
+		minNumAttempts  int
+		maxNumAttempts  int
+		expectedErrText string
+	}{
+		{
+			desc: "succeeds when no errors are ever given",
+			ctx:  context.Background(),
+			opts: Options{
+				InitialBackoff: time.Microsecond * 10,
+				MaxBackoff:     time.Second,
+				Multiplier:     2,
+				MaxRetries:     1,
+			},
+			retryFunc:   func() error { return nil },
+			maxAttempts: 3,
+
+			minNumAttempts: 0,
+			maxNumAttempts: 0,
+		},
+		{
+			desc: "succeeds after one faked error",
+			ctx:  context.Background(),
+			opts: Options{
+				InitialBackoff: time.Microsecond * 10,
+				MaxBackoff:     time.Second,
+				Multiplier:     2,
+				MaxRetries:     1,
+			},
+			retryFunc:   errorUntilAttemptNumFunc(1),
+			maxAttempts: 3,
+
+			minNumAttempts: 1,
+			maxNumAttempts: 1,
+		},
+		{
+			desc: "errors when max attempts is exhausted",
+			ctx:  context.Background(),
+			opts: Options{
+				InitialBackoff: time.Microsecond * 10,
+				MaxBackoff:     time.Second,
```

Comment:
I would recommend reducing the backoff time - you don't need a full second of testing to know the test is successful.

Reply:
fixed!